### PR TITLE
Implement users

### LINKED ISSUE #83: Associate deadlines to signed-in account
Currently Firebase stores a single global collection of deadlines shared by all instances of the application, regardless of the signed-in account, if any. We should integrate the Firebase authentication to the deadline repository so that each user has their own list of deadlines, and cannot access the deadlines of other users.

### PR DESCRIPTION
**Pull request by felko**
This PR implements users to the application. Deadlines are now associated to an user ID.
A non authenticated user can still use the application, thanks to the anonymous users feature of Firebase.
At the first start of the application, an anonymous user is created. When the user signs in with a google account, all deadlines are associated to the new user.
When logging out, a new anonymous user is created. This lets us assume in the entire codebase that thee's always an user to which we can associate the deadlines.

Closes #83, #125

### CODE REVIEW

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/MainActivityTest.kt [hunk 1]
```diff
@@ -237,5 +240,14 @@ class MainActivityTest {
...
                 )
             )
+
+        @Singleton
+        @Provides
+        fun provideAuthRepository(): AuthRepository =
+            MockAuthRepository()
+    }
+
+    companion object {
```
Comment (Matathiasq13824) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/MainActivityTest.kt:250]: I'm not really sure. Why this need to be in a companion object ?

File: app/src/main/java/com/github/multimatum_team/multimatum/ProfileActivity.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.github.multimatum_team.multimatum.model.User
+import com.github.multimatum_team.multimatum.viewmodel.UserViewModel
+import dagger.hilt.android.AndroidEntryPoint
+
+/**
+ * A simple [Fragment] subclass.
+ * Use the [ProfileActivity.newInstance] factory method to
+ * create an instance of this fragment.
+ */
+@AndroidEntryPoint
```
Comment (Matathiasq13824) [inline app/src/main/java/com/github/multimatum_team/multimatum/ProfileActivity.kt:19]: You could change this comment to explain briefly what this fragment do

File: app/src/main/java/com/github/multimatum_team/multimatum/SignInActivity.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import com.firebase.ui.auth.AuthUI
+import com.firebase.ui.auth.FirebaseAuthUIActivityResultContract
+import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult
+import com.google.firebase.auth.FirebaseAuth
+import dagger.hilt.android.AndroidEntryPoint
+
+@AndroidEntryPoint
+class SignInActivity : AppCompatActivity() {
```
Comment (Matathiasq13824) [inline app/src/main/java/com/github/multimatum_team/multimatum/SignInActivity.kt:16]: You could add comments to describe the class and the functions

**Review by Matathiasq13824 — CHANGES_REQUESTED** [review]
Great job, it was a collosal work ! However please next time try to separate it on multiple pr (35 files is quite a lot ^^')
I have only put some minor comments. However, it could be nice to add some tests to increase coverage in the next sprint. When the CI has finished, the test pass and some minor modification are made, I will approve

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/MainActivityTest.kt (hunk 1 above)
Comment (felko) [reply app/src/androidTest/java/com/github/multimatum_team/multimatum/MainActivityTest.kt:250]: Before I moved this test to the roboelectric tests, the mockSharedPreferences was taken from the MainSettingsActivityTest, which I kept in the instrumented tests. Therefore I simply copied what was in MainSettingsActivityTest and put it in MainActivityTest.

**Review by macintosh1398 — APPROVED** [review]
What a great work, a lot of new code but the modularity is good and the code is clean. The previous comment is relevant, so nothing to add. Maybe just focus on the tests next week.

**Review by Matathiasq13824 — APPROVED** [review]

**Pull request merged**